dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +199,301 @@ cdef int _local_tracker(
+    pft_max_trials : int
+        Maximum number of trials for the particle filtering tractography
+        (Prevents infinite loops).
+    particle_count : int
+        Number of particles to use in the particle filter.
+    particle_paths : array, float, 4d, (2, particle_count, pft_max_steps, 3)
+        Temporary array for paths followed by all particles.
+    particle_dirs : array, float, 4d, (2, particle_count, pft_max_steps, 3)
+        Temporary array for directions followed by particles.
+    particle_weights : array, float, 2d, (2, particle_count)
+        Temporary array for the weights of particles.
+    particle_states : array, float, (2, particle_count, 2)
+        Temporary array for the states of particles.
+
+    Returns
+    -------
+    end : int
+        Length of the tracked streamline
+    tissue_class : TissueClass
+        Ending state of the streamlines as determined by the TissueClassifier.
+
+    """
+    cdef:
+        int i
+        TissueClass tissue_class
+        double dir[3]
+        double vs[3]
+        double seed[3]
+
+    if (seed_pos.shape[0] != 3 or first_step.shape[0] != 3 or
+            voxel_size.shape[0] != 3 or streamline.shape[1] != 3):
+        raise ValueError('Invalid input parameter dimensions.')
+
+    for i in range(3):
+        dir[i] = first_step[i]
+        vs[i] = voxel_size[i]
+        seed[i] = seed_pos[i]
+
+    i = _pft_tracker(dg, tc, seed, dir, vs, streamline,
+                     directions, step_size, &tissue_class, pft_nbr_back_steps,
+                     pft_max_steps, pft_max_trials, particle_count,
+                     particle_paths, particle_dirs, particle_weights,
+                     particle_states)
+    return i, tissue_class
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef _pft_tracker(DirectionGetter dg,
+                  ConstrainedTissueClassifier tc,
+                  double* seed,
+                  double* dir,
+                  double* voxel_size,
+                  np.float_t[:, :] streamline,
+                  np.float_t[:, :] directions,
+                  double step_size,
+                  TissueClass * tissue_class,
+                  int pft_nbr_back_steps,
+                  int pft_max_steps,
+                  int pft_max_trials,
+                  int particle_count,
+                  np.float_t[:, :, :, :] particle_paths,
+                  np.float_t[:, :, :, :] particle_dirs,
+                  np.float_t[:, :] particle_weights,
+                  np.int_t[:, :, :] particle_states):
+    cdef:
+        int i, pft_trial, pft_streamline_i, pft_nbr_steps, strl_array_len
+        double point[3]
+        double voxdir[3]
+        void (*step)(double* , double*, double) nogil
+
+    copypoint(seed, point)
+    copypoint(seed, &streamline[0,0])
+    copypoint(dir, &directions[0, 0])
+
+    tissue_class[0] = TRACKPOINT
+    pft_trial = 0
+    i = 0
```

Comment:
I have a hard time following `i`. If I understand correctly, It represents the real number of points in `streamline` (and not `len(streamline)` which never change). In that case, why is it not starting at `1`? Would it be simpler instead of all the `i += 1; break;`?

Lets say `get_direction` returns False (which is good, wtf?), we add the current point and i++, then ``check_point` returns `ENDPOINT`, i++ again and break. This is strange for the reader to see another i++ here because you haven't added a new point.

Reply:
Here. `i+=1` is done when we want to keep the last point (`tissue_class == ENDPOINT or INVALIDPOINT`), not when `tissue_class == OUTSIDEIMAGE`.  I tried to clarify the code. Thx.

For the `get_direction(.)`, It is a function name issue here. It should be called `update_direction(point, direction)`, returning 0 if there is no error, not 0 otherwise. With the current name, I would rather be having it returning True when the get_direction succeed. We can open an issue for this to be done in an other PR.